# Programación Lineal

## Dieta Optima

### Planteamiento del problema:

Se requiere diseñar una mezcla cuya combinación de alimentos tenga el menor costo posible y que al mismo tiempo, cubra los requerimientos nutritivos para pollos machos de 22 a 33 días de edad con un aporte de: 3.100 kcal, 19.25% de proteína, 0.817% de calcio, 0.408% de Fósforo disponible. Para obtener un resultado que permita generalizar los porcentajes de cada alimento la cantidad a calcular será de 100 Kgr.  

Se tienen los siguientes ingredientes alimenticios:
    
Alimento           |Precio(p/kg)|EM (kcal/kg)|P.B.(%)|Calcio(%)|Fosforo Disp.(%)|Cantidad
----|----
Maíz               |	2	    |3.381	     |8.26	 |0.11	   |0.08            |$X_1$
Soya, Aceite       |	9	    |8.79	     |0	     |0	       |0               |$X_2$
Soya Harina        |	3	    |2.256	     |45.3	 |0.24	   |0.18            |$X_3$
Soya Integral      |	3.5	    |3.281	     |37	 |0.23	   |0.17            |$X_4$
Carbonato Calcitico|	3	    |0	         |0	     |38.4	   |0               |$X_5$
Fosfato Bicalcico  |	8	    |0	         |0	     |25.5	   |18.5            |$X_6$
$Requerimientos$   |    $min$   |$3.1$       |$19.25$|$0.817$  |$0.408$         |$100$



In [19]:
#librerias
from pulp import *

#Crear Problema
prob=LpProblem("Diseño Optimo de Dieta: Costo Mínimo", LpMinimize)

#Crear vector de alimentos:
alimento=["Maíz","Soya,Aceite","Soya Harina","Soya Integral","Carbonato Calcitico","Fosfato Bicalcico"]

#Crear vector de requerimientos:
requerimiento=[3.1,19.25,0.817,0.408,1]

#Crear diccionarios con la información:
precio={alimento[0]:2.0,
        alimento[1]:9.0,
        alimento[2]:3.0,
        alimento[3]:3.5,
        alimento[4]:3.0,
        alimento[5]:8.0}

EM={alimento[0]:3.381,
    alimento[1]:8.79,
    alimento[2]:2.256,
    alimento[3]:3.281,
    alimento[4]:0,
    alimento[5]:0}

PB={alimento[0]:8.26,
    alimento[1]:0,
    alimento[2]:45.3,
    alimento[3]:37,
    alimento[4]:0,
    alimento[5]:0}

calcio={alimento[0]:0.11,
        alimento[1]:0,
        alimento[2]:0.24,
        alimento[3]:0.23,
        alimento[4]:38.4,
        alimento[5]:25.5}

fosforo={alimento[0]:0.08,
         alimento[1]:0,
         alimento[2]:0.18,
         alimento[3]:0.17,
         alimento[4]:0,
         alimento[5]:18.5}

#Crear las variables
variables= LpVariable.dicts("Alimentos", alimento,0,None)

#Se agrega la funcion objetivo

prob+= lpSum ([precio[i]*variables[i] for i in alimento])

#Constraints are added:

prob+= lpSum([EM[i]*variables[i] for i in alimento]) == requerimiento[0]
prob+= lpSum([PB[i]*variables[i] for i in alimento]) == requerimiento[1]
prob+= lpSum([calcio[i]*variables[i] for i in alimento]) == requerimiento[2]
prob+= lpSum([fosforo[i]*variables[i] for i in alimento]) == requerimiento[3]
prob+= lpSum([1*variables[i] for i in alimento]) == requerimiento[4]

#Solution to the problem: 
prob.solve()
print ("Status:",LpStatus[prob.status])
print("")
for v in prob.variables():
    print (v.name,"=",100*v.varValue,"%")
print("")
print("Costo =",value(prob.objective))

Status: Optimal

Alimentos_Carbonato_Calcitico = 0.66899911 %
Alimentos_Fosfato_Bicalcico = 1.6079352000000002 %
Alimentos_Maíz = 63.61654 %
Alimentos_Soya,Aceite = 0.0 %
Alimentos_Soya_Harina = 16.576617 %
Alimentos_Soya_Integral = 17.529909 %

Costo = 2.5318809143


El costo mínimo del alimento es de $\$$2.53 con una cantidad de 63.61$\%$ de Maíz, 16.57$\%$ de Soya Harina, 17.53$\%$ de Soya Integral, 0.66$\%$ de carbonato calcitico y 1.6$\%$ de fosfato bicalcico. Es importante notar que se satisfacen todas las restricciones. Así mismo, es sencillo cambiar los precios respectivos de los productos como las restricciones dadas por lo que se puede tomar como un archivo general. 

__Reference__:  
Huici, R. (2012). “M.S. Excel Aplicado a la Formulación de alimentos para nutrición animal”. Bolivia: n.a, Primera edición. 